### ISSUE TITLE
Houdini 17.5 and USD Hydra plugin?

### ISSUE THREAD

**Issue by DanEnglesson** [dreamworksanimation/dwa_usd_plugins#3]
Hi,

I've tested out building the Hydra plugin using Houdini 17.5.327 and I got pretty close but seems like there is a few things that I can't seem to figure out and just throwing it out there to see if you have solved this. Please feel free to reject this issue.

First off, houdini 17.5 now uses a USD that ships with Houdini, and building anything using anything other than it's inhouse USD will make it pretty sad. So for instance, building the OP_gusd that ships with Houdini is now done like this:

```
export HOME=/usr/local/USD.19
cmake3 /opt/hfs17.5.327/toolkit/usd_houdini_plugins -DCMAKE_C_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/gcc -DCMAKE_CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/g++ -DCMAKE_PREFIX_PATH=/opt/hfs17.5.327/toolkit
make -j12 install
```

I've looked at how that is setup and used that as a base to generate this [CMakeLists.txt](https://github.com/dreamworksanimation/dwa_usd_plugins/files/4307902/CMakeLists.txt) for the Hydra plugin. It compiles fine but segfaults om some missing functions that are not dynamically linked. First test it segfaults on startup with this:

```
/opt/hfs17.5.327/bin/houdinicore-bin: symbol lookup error: /usr/local/USD.19/houdini17.5/dso/Hydra.so: undefined symbol: _ZN32pxrInternal_v0_19__pxrReserved__16GusdGU_PackedUSD7installER19GA_PrimitiveFactory
```
That points to:
pxrInternal_v0_19__pxrReserved__::GusdGU_PackedUSD::install(GA_PrimitiveFactory&)
which by inspecting the .so file is undefined in the Hydra.so file. It does exists in the libHoudiniUSD.so though file, but it seems to fall off during dynamically linking it.

It fails here:
https://github.com/dreamworksanimation/dwa_usd_plugins/blob/master/third_party/houdini/plugin/Hydra/plugin.cpp#L34

Removing that line made it go further since that is done on startup in 17.5 and above when talking to sidefx, so no need to do that anymore. But there are still some functions that are missing and it now fails here:
```
/opt/hfs17.5.327/bin/houdinicore-bin: symbol lookup error: /usr/local/USD.19/houdini17.5/dso/Hydra.so: undefined symbol: _ZNK32pxrInternal_v0_19__pxrReserved__16GusdGU_PackedUSD10getUsdPrimE16UT_ErrorSeverity
```
looking at the libgusd.so file that is the lib it used before the libHoudiniUSD.so file, it does seem to have more dynamic functions exposed.
```
nm -D /usr/local/USD.19/houdini17.5/lib/libgusd.so | grep getUsdPrim
000000000014ab40 W _ZNK32pxrInternal_v0_19__pxrReserved__15GusdMeshWrapper10getUsdPrimEv
00000000002882b0 T _ZNK32pxrInternal_v0_19__pxrReserved__16GusdGU_PackedUSD10getUsdPrimE16UT_ErrorSeverity
```
vs libHoudiniUSD.so
```
nm -D /opt/hfs17.5.327/dsolib/libHoudiniUSD.so | grep getUsdPrim
0000000000460cf0 T _ZN32pxrInternal_v0_19__pxrReserved__23HUSDgetUsdPrimPredicateE25HUSD_PrimTraversalDemands
```
I'm not too familiar with what would make the ligbusd.so file have that function show up vs the libHoudiniUSD.so file not having it. I take it that if I were to add the USD_API to that function that it would make it to the dynamically linked functions, but that would require rebuilding libHoudiniUSD.so I think.

If you have any pointers or ideas of solving this they would be more than welcome.
Many thanks,
Dan

**Comment by ablev-dwa** [comment]
Hi Dan.

I'm afraid I don't have any helpful advice for you here. We have skipped Houdini 17.5 in favor of Houdini 18, so that we can use [HoudiniUsdBridge](https://github.com/sideeffects/HoudiniUsdBridge) to use Houdini with our internal USD build. We haven't yet tried building this extension with Houdini 18, but if we attempt it, I expect we may run into the same problems you are with API changes to gusd. I'm hoping we won't need to do that though, since Houdini 18 has some pretty great Hydra integration already, which I think removes the need for this extension.

**Comment by DanEnglesson** [comment]
Hi Alan,

Yeah the HoudiniUsdBridge will be the thing to use in 18, and most likely pave the way to build this plugin as well. Although as you mentioned, maybe there will not be much need for this extension in 18 with Lops etc. However, this plugin will still be needed if you were to take it outside of the lops workflow since hydra is not working with SOPs, so we would need to compile this too. But with the HoudiniUsdBridge it might be easier to get going.

To fix this for 18, if needed, I think the main thing to do is to make certain functions having the USD_API added to them to be exposed in the dynamic linking.

I think we will do the same as you guys and target 18 instead of trying to fix this for 17.5.
Closing this issue.

Thanks for your help Alan.
Cheers,
Dan
